samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.samza.sql.interfaces.RelSchemaConverter;
+import org.apache.samza.sql.interfaces.RelToSamzaMsgConverter;
+import org.apache.samza.sql.interfaces.SamzaToRelMsgConverter;
+import org.apache.samza.sql.interfaces.SourceSchemaResolver;
+import org.apache.samza.sql.interfaces.SystemStreamConfig;
+import org.apache.samza.sql.planner.QueryPlanner;
+import org.apache.samza.sql.runner.SamzaSqlApplicationConfig;
+import org.apache.samza.sql.testutil.SamzaSqlQueryParser;
+
+
+/**
+ * This class is used to populate the StreamGraph using the SQL queries.
+ * This class contains the core of the SamzaSQL control code that converts the SQL statements to calcite relational graph.
+ * It then walks the relational graph and then populates the Samza's {@link StreamGraph} accordingly.
+ */
+public class QueryTranslator {
+
+  private final Map<String, Optional<SamzaToRelMsgConverter>> samzaToRelMsgConvertersBySystem;
+  private final ScanTranslator scanTranslator;
+  private final Map<String, RelSchemaConverter> schemaConvertersBySystem;
+  private final Map<String, SourceSchemaResolver> schemaResolversBySystem;
+  private final Map<String, Optional<RelToSamzaMsgConverter>> relToSamzaMsgConvertersBySystem;
+  private final SamzaSqlApplicationConfig sqlConfig;
+
+  public QueryTranslator(SamzaSqlApplicationConfig sqlConfig) {
+    this.sqlConfig = sqlConfig;
+    schemaResolversBySystem = sqlConfig.getSchemaResolversBySystem();
+    samzaToRelMsgConvertersBySystem = sqlConfig.getSamzaToRelMsgConvertersBySystem();
+    relToSamzaMsgConvertersBySystem = sqlConfig.getRelToSamzaMsgConvertersBySystem();
+    schemaConvertersBySystem = sqlConfig.getSchemaConvertersBySystem();
+    scanTranslator = new ScanTranslator(samzaToRelMsgConvertersBySystem);
+  }
+
+  public void translate(SamzaSqlQueryParser.QueryInfo queryInfo, StreamGraph streamGraph) {
+    QueryPlanner planner = new QueryPlanner(schemaResolversBySystem, schemaConvertersBySystem,
+        sqlConfig.getInputSystemStreamConfigBySource(), sqlConfig.getUdfMetadata());
+    final RelRoot relRoot = planner.plan(queryInfo.getSelectQuery());
+    final TranslatorContext context = new TranslatorContext(streamGraph, relRoot);
+    final RelNode node = relRoot.project();
+
+    node.accept(new RelShuttleImpl() {
+      @Override
+      public RelNode visit(TableScan scan) {
+        RelNode node = super.visit(scan);
+        scanTranslator.translate(scan, context);
+        return node;
+      }
+
+      @Override
+      public RelNode visit(LogicalFilter filter) {
+        RelNode node = visitChild(filter, 0, filter.getInput());
+        new FilterTranslator().translate(filter, context);
+        return node;
+      }
+
+      @Override
+      public RelNode visit(LogicalProject project) {
+        RelNode node = super.visit(project);
+        new ProjectTranslator().translate(project, context);
+        return node;
+      }
+    });
+
+    SystemStreamConfig outputSystemConfig =
```

Comment:
I would expect the output should be translated from "insert into" (TableAlter relNode?) to sendTo() operator.

Reply:
I think that it is assumed to be *always* sendTo, since we are only dealing with "insert into". Hence, it is directly coded in line 98 below instead of going through a translator. I am OK w/ it now, w/ some additional documentation to explain this implementation here.